### PR TITLE
Add method for disjoint tracking and classification

### PR DESCRIPTION
**Pull request by oharrald-Dstl**
These changes introduce a simple method of classifying a target in parallel with tracking its kinematics.
An implementation of a hidden Markov model is used to provide the classifying-side of this process, whereby successive indirect observations of the target's hidden class are taken in order to attain a categorical distribution over the (discrete) set of possible hidden classes.
The method is 'disjoint', in that the target dynamics are not used to attain an understanding of the classification of the target, and vice versa. The two processes are separate, and are only combined during the data association phase of the tracking loop, whereby the likelihood of a particular detection being associated to a track takes both the track classification and kinematics in to account.
Composite state types are introduced to accommodate this. For example, if we are interested in tracking the kinematics of the target (using a Kalman filter), and classifying the target (using the implemented HMM tracking modules), we initiate with a prior composite state, composed of both a Gaussian state and categorical state.

- [x] Add a sphinx gallery tutorial
- [ ] Increase code coverage to ~100% for all modified/added files
- [ ] Remove composite sensor

### CODE REVIEW

File: stonesoup/hypothesiser/composite.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        : :class:`~.MultipleHypothesis`
+            A container of :class:`~CompositeHypothesis` objects, each of which containing a
+            sequence of :class:`~.SingleHypothesis` objects
+
+        Notes
+        ----
+        * Detections are required to have a timestamp, in order for the composite null-hypothesis
+        to have a timestamp.
+        * A 'weightings' attribute should be implemented to adjust sub-hypotheses' contributions
+        to the composite hypothesis weights.
```
Comment (sdhiscocks) [inline stonesoup/hypothesiser/composite.py:68]: Warning from docs build:
```
stonesoup/hypothesiser/composite.py:docstring of stonesoup.hypothesiser.composite.CompositeHypothesiser.hypothesise:25: WARNING: Bullet list ends without a blank line; unexpected unindent.
```

**Comment by jmbarr** [comment]
3000+ new lines eh? This might take a while. Is there an example notebook that might assist in testing?

**Comment by oharrald-Dstl** [comment]
> 3000+ new lines eh? This might take a while. Is there an example notebook that might assist in testing?

I'll be adding a sphinx-gallery tutorial on how to use most of the components introduced. Hopefully the notebook version of that will be useful.

File: stonesoup/hypothesiser/composite.py (hunk 1 above)
Comment (oharrald-Dstl) [reply stonesoup/hypothesiser/composite.py:68]: I think this is now fixed. Though there is now another docs build issue.

File: stonesoup/initiator/composite.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+            for i, sub_initiator in enumerate(self.sub_initiators):
+
+                try:
+                    # Check if detection has sub-detection for this state index
+                    detection_index = mapping.index(i)
+
+                    # Get sub-detection and initiate a (sub)track with it
+                    sub_detection = detection[detection_index]
+                    sub_tracks = sub_initiator.initiate({sub_detection}, timestamp=timestamp)
```
Comment (sdhiscocks) [inline stonesoup/initiator/composite.py:67]: I'm wondering if you should group all relevant detections together, and therefore call each sub initiator once. Initiation approach may be somehow reliant on the assumption that all detections from one sensor at the same timestamp are present.

**Review by sdhiscocks — COMMENTED** [review]
Overall looks really good :+1:

I've focussed my review on composite element, and overall code.

I still need to look at the example more closely, and think we need input from those familiar with HMM to review the categorical elements more closely (models, predictor/updater).

File: stonesoup/hypothesiser/composite.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+        # as each detection is composite, it will have a set of sub-hypotheses paired to it
+        detections_hypotheses = defaultdict(list)
+
+        try:
+            # get timestamp from one of the detections
+            timestamp = next(iter(detections)).timestamp
+        except StopIteration:
+            # nothing to iterate over => no detections => return an empty set of hypotheses
+            return MultipleHypothesis({}, normalise=True, total_weight=1)
```
[the remark this answers is not in the document]
Comment (oharrald-Dstl) [reply stonesoup/hypothesiser/composite.py:87]: Have removed this code as its not needed.

File: stonesoup/initiator/composite.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                    # from the sub-detection
+                    # Instead initiate sub-state as the sub-initiator's prior
+                    prior = self._prior_state[i]
+                    states.append(prior)
+
+                    # Add missed detection hypothesis to composite hypothesis
+                    hypotheses.append(
+                        SingleHypothesis(None, MissedDetection(timestamp=detection.timestamp)))
+                else:
+                    update = track[0]  # Get first state of track
```
[the remark this answers is not in the document]
Comment (oharrald-Dstl) [reply stonesoup/initiator/composite.py:81]: Last, good point.

File: stonesoup/types/detection.py [hunk 5]
```diff
@@ -60,3 +60,43 @@ def __init__(self, state_vector=None, *args, **kwargs):
...
+        metadata = dict()
+        for sub_detection in self.sub_states:
+            metadata.update(sub_detection.metadata)
+        return metadata
+
+
+class CompositeMissedDetection(CompositeDetection):
+
+    def __bool__(self):
+        return False
```
[the remark this answers is not in the document]
Comment (oharrald-Dstl) [reply stonesoup/types/detection.py:102]: Think I need something similar for the composite state type too.

File: stonesoup/types/state.py [hunk 6]
```diff
@@ -326,4 +328,138 @@ def covar(self):
...
+    def append(self, value):
+        """Add value at end of :attr:`sub_states`.
+
+        Parameters
+        ----------
+        value: State
+            A state object to be added at the end of :attr:`sub_states`.
+        """
+        self.sub_states.append(value)
+        self._check_timestamp()
```
[the remark this answers is not in the document]
Comment (oharrald-Dstl) [reply stonesoup/types/state.py:465]: The only change I'd need to make would be situations in which I've created composite states through looping, which can be redone.
I'm all for that. Would there ever be a situation in which modifying a composite state would be useful though?

File: stonesoup/initiator/composite.py (hunk 2 above)
Comment (oharrald-Dstl) [reply stonesoup/initiator/composite.py:67]: Agreed. That would look similar then to the composite hypothesiser method.

File: stonesoup/types/state.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sdhiscocks) [reply stonesoup/types/state.py:465]: > The only change I'd need to make would be situations in which I've created composite states through looping, which can be redone.

I did look, and I don't think you ever do this in the main code. Possibly in tests, but would have to look again.

> I'm all for that. Would there ever be a situation in which modifying a composite state would be useful though?

I don't think so. Once you modify it, it isn't the same state anymore, so can only think you'd do this on creation; and you can just use a list, then create the state (which is what the main code seems to do in all cases).

**Comment by jmbarr** [comment]
> Overall looks really good 👍
> 
> I've focussed my review on composite element, and overall code.
> 
> I still need to look at the example more closely, and think we need input from those familiar with HMM to review the categorical elements more closely (models, predictor/updater).

I think we need a separate pull request for this (see below)

**Comment by jmbarr** [comment]
An impressive piece of work, but one which needs to be split several ways so that it can be interrogated properly. My impressions/suggestions:

1. Separate the categorical classes and submit a pull request for those. They are a necessary precursor to the (dis)joint tracking but have more widespread application so it's worth getting them out into the world and not having this PR hold them up.
2. I don't like `CompositeState`. It seems to subvert the simplicity of Stone Soup being based around `State` and create a whole set of parallel classes. We may need to augment `State` to allow it to work with labels or perhaps with StateVectors of varying type but fundamentally these are still states. 
3. (Related) there's now a whole ecosystem of CompositeXXXX classes which adds (I think unnecessary) complexity. I don't think it's a bad idea to allow others to create composite states for their own purposes and perhaps demonstrate how to do that in an example, but I think having composite versions of large numbers of Stone Soup classes in the codebase isn't a good idea.
4. Why 'disjoint'? Is it because kinematics and categories are kept separate? What would 'joint' look like?

**Comment by oharrald-Dstl** [comment]
> An impressive piece of work, but one which needs to be split several ways so that it can be interrogated properly. My impressions/suggestions:
> 
> 1. Separate the categorical classes and submit a pull request for those. They are a necessary precursor to the (dis)joint tracking but have more widespread application so it's worth getting them out into the world and not having this PR hold them up.
> 2. I don't like `CompositeState`. It seems to subvert the simplicity of Stone Soup being based around `State` and create a whole set of parallel classes. We may need to augment `State` to allow it to work with labels or perhaps with StateVectors of varying type but fundamentally these are still states.
> 3. (Related) there's now a whole ecosystem of CompositeXXXX classes which adds (I think unnecessary) complexity. I don't think it's a bad idea to allow others to create composite states for their own purposes and perhaps demonstrate how to do that in an example, but I think having composite versions of large numbers of Stone Soup classes in the codebase isn't a good idea.
> 4. Why 'disjoint'? Is it because kinematics and categories are kept separate? What would 'joint' look like?

1. I agree
2. I'm not sure what you mean.
3. I've added the necessary composite components for tracking in Stone Soup. I can see the CompositePredictor being redundant in many places. Which others?
4. I would consider a 'joint' approach to be direct implication of classification from dynamics, and vice versa. Eg. dynamics: changing direction quickly -> category: bicycle, dynamics: slow, stopping often -> category: bus.
Where there is a relationship between classification and dynamics given in the association step, 'disjoint' is probably not the best title for this work.

**Comment by jmbarr** [comment]
> > An impressive piece of work, but one which needs to be split several ways so that it can be interrogated properly. My impressions/suggestions:
> > 
> > 1. Separate the categorical classes and submit a pull request for those. They are a necessary precursor to the (dis)joint tracking but have more widespread application so it's worth getting them out into the world and not having this PR hold them up.
> > 2. I don't like `CompositeState`. It seems to subvert the simplicity of Stone Soup being based around `State` and create a whole set of parallel classes. We may need to augment `State` to allow it to work with labels or perhaps with StateVectors of varying type but fundamentally these are still states.
> > 3. (Related) there's now a whole ecosystem of CompositeXXXX classes which adds (I think unnecessary) complexity. I don't think it's a bad idea to allow others to create composite states for their own purposes and perhaps demonstrate how to do that in an example, but I think having composite versions of large numbers of Stone Soup classes in the codebase isn't a good idea.
> > 4. Why 'disjoint'? Is it because kinematics and categories are kept separate? What would 'joint' look like?
> 
> 1. I agree
> 2. I'm not sure what you mean.
> 3. I've added the necessary composite components for tracking in Stone Soup. I can see the CompositePredictor being redundant in many places. Which others?
> 4. I would consider a 'joint' approach to be direct implication of classification from dynamics, and vice versa. Eg. dynamics: changing direction quickly -> category: bicycle, dynamics: slow, stopping often -> category: bus.
>    Where there is a relationship between classification and dynamics given in the association step, 'disjoint' is probably not the best title for this work.

1. Excellent
2. I think the barrier to creation of new classes, particularly ones that don't inherit, should be high. That keeps the code cleaner and simpler. So before we create `CompositeState` (or any `CompositeXlass`) we should try hard to make the method work with `State`, or at least something derived from `State`.
3. So that's an example of complexity arising from the choice to create a new set of states, `CompositeState`, `CompositePrediction`, CompositeDetection`, etc, and for the most part that should be reserved for where necessary. If you can see redundancy, don't introduce it. Also, I don't think a `CompositeDetection` can be a thing. A detector delivers a detection - if it's a combination of kinematics and classes then that's what it is. Still a detection. Further, I think the mapping= keyword would take care of many issues with choosing which parameters are appropriate for the more 'functional' classes (e.g. Predictors/Associators/Updaters).
4. Understood - I think. The kinematic predict and update steps are separate from the classifier predict and update steps and only 'meet' in the associator, where a combined probability of association is used (is that right?). This begs an interesting question. Can you set up an experiment where you compare this version against one where there is a truly 'joint' predict and update? (And can you write it up for next year's Fusion conference?)

**Pull request closed without merging**